### PR TITLE
Edits to the Map Sorting proposal

### PR DESCRIPTION
**Pull request by calda**
I've made some edits to the proposal! Let me know what you think.

### CODE REVIEW

File: proposals/0000-map-sorting.md [hunk 1]
```diff
@@ -117,9 +122,22 @@ extension MutableCollection where Self: RandomAccessCollection {
 
 This is an ABI-compatible addition with no impact on source compatibility.
 
+## Future Directions
+
+### Provide `<` as the default `areInIncreasingOrder` predicate when `Value: Comparable`
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md:129]: As the actual predicate, not the default, right? We don't need a predicate parameter at all for `Value: Comparable`.

File: proposals/0000-map-sorting.md [hunk 2]
```diff
@@ -117,9 +122,22 @@ extension MutableCollection where Self: RandomAccessCollection {
...
+
+A future addition to the Standard Library's sorting API could provide a version of `sort(on:)` and `sorted(on:)` where `<` is provided as the default sorting predicate for `Comparable` values:
+
+```swift
+people.sort(on: { $0.age })
+people.sort(on: \.age)
+```
+
+This follows the precedent set by the existing `sort(by:)` and `sorted(by:)` methods. The authors chose to exclude this additional overload from this initial proposal, as it is purely additive and could be included in either a later proposal or a final stage of this proposal (pending feedback during the review process).
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: The authors chose to exclude <s>this additional overload</s> these overloads from...?

File: proposals/0000-map-sorting.md [hunk 3]
```diff
@@ -117,9 +122,22 @@ extension MutableCollection where Self: RandomAccessCollection {
...
+
 ## Alternatives considered
 
-### Argument label naming  
+### Alternative argument labels
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: 'Alternative' is redundant due to the superior section name.

File: proposals/0000-map-sorting.md [hunk 4]
```diff
@@ -130,18 +148,10 @@ This is an ABI-compatible addition with no impact on source compatibility.
 The `by` label is a perfect candidate to describe the metric used to sort the sequence. `using`, on its turn, is just as fitting for a predicate or an operator. The pair in question is perhaps the only one that always boils down to a proper
 sentence - «*Sort(ed) **by** a metric **using** a predicate*». Nevertheless, the author is convinced in the superior importance of preserving API uniformity and consistency with existing API the Standard Library developers have worked so hard to keep. With ABI Stability kicking in, we no longer have the opportunity for amendments to public API signatures and must be especially careful in this regard.
 
-### Convenience overloads
+### Alternative argument order
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: Here too.

File: proposals/0000-map-sorting.md [hunk 5]
```diff
@@ -130,18 +148,10 @@ This is an ABI-compatible addition with no impact on source compatibility.
...
 
-It has been considered to mirror the `sort()` & `sorted()` precedent by implementing two additional overloads that specialize on ascending order. The latter is known to be by far the most common scenario in practice; abundant enough for the aforementioned API to have overcome the moratorium on trivially composable sugar. [Browsing](https://forums.swift.org/t/map-sorting/21421/20?u=anthonylatsis) the [Swift Source Compatibility Suite](https://github.com/apple/swift-source-compat-suite) shows a 9:1 ratio in favor of parameter-less sorting method usage. At the same time, the presence or absence of a parameter becomes less tangible as the number of parameters increases.
+It could be desirable to rearrange the arguments so the `transform` closure follows the `areInInreasingOrder` predicate. This would allow the caller to make use of trailing-closure syntax:
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: We have a third argument though - the Schwartzian transform. A closure must come last to enable trailing closure syntax.

File: proposals/0000-map-sorting.md [hunk 6]
```diff
@@ -34,17 +34,21 @@ people.sort { $0.age < $1.age }
...
 
 ## Proposed solution
 
-Add an overload for both the nonmutating `sorted` and in-place `sort` methods on `Sequence` and `MutableCollection` respectively. A mapping closure on `Element` will lead the argument list, followed by the well known `areInIncreasingOrder` predicate and, finally, a flag for opting into the already mentioned [Schwartzian Transform](https://en.wikipedia.org/wiki/Schwartzian_transform) optimization. `transform` is deliberately positioned before the predicate to respect logical and type-checking order. Here are some example usages:
+The authors propose to add an overload for both the nonmutating `sorted` and in-place `sort` methods on `Sequence` and `MutableCollection` respectively. A mapping closure `(Element) -> Value` will lead the argument list, followed by the well known `areInIncreasingOrder` predicate of the type `(Value, Value) -> Bool`. Additionally, we propose a flag for opting into the already mentioned [Schwartzian Transform](https://en.wikipedia.org/wiki/Schwartzian_transform) optimization.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: The type of the mapping closure is a good insight, but mentioning the type of `areInIncreasingOrder` as well seems like an overkill. Maybe just clarify with 'comparison predicate'? 
___
Since you already provide a link not far above, can we remove this second link to Wikipedia?

File: proposals/0000-map-sorting.md [hunk 7]
```diff
@@ -8,13 +8,13 @@
 
 ## Introduction
 
-This proposal presents an addition to the Standard Library in an effort to make map sorting and, eventually, closureless key-path sorting a functional prerequisite for Swift.
+The authors propose adding an additional `sort` and `sorted` method on `MutableCollection` and `Sequence`, respectively, to improve the ergonomics for common use cases of the existing `sort` and `sorted` APIs.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: Can we keep the second half of the sentence to be more explicit about the purpose of the addition?

Nit: `MutableCollection` and `Sequence` ->  `Sequence` and `MutableCollection`.
Nit:  Rephrase 'adding an additional'?

File: proposals/0000-map-sorting.md [hunk 8]
```diff
@@ -34,17 +34,21 @@ people.sort { $0.age < $1.age }
 chats.sort { $0.lastMsg.date > $1.lastMsg.date }
 ```
 
-Most often, however, all we need to determine sorting is a key-path. Other times, a comparator on its own becomes inconvenient or inefficient.
-* The `$0.property < $1.property` syntax often leads to copy-and-paste bugs.
-* The base metric is not always trivial to obtain, in which case a predicate alone instigates duplication and makes it harder to grasp the sorting order.
-* When the values are expensive to retrieve, the predicate obscuring the comparison also becomes a major obstacle for optimizations, such as trading memory for speed to warrant that each value is computed once per element rather than O(*n* log*n*) times. To clarity the impact, applying the latter to a ϴ(*n*) operation theoretically speeds up sorting by a factor of 10 for an array of 1000 elements. 
-Thereby, the goal is to introduce an API that decouples the comparison of values from their calculations, favoring optimizations and bringing ergonomic benefits for an ample range of cases.
+In many circumstances, this syntax can cause issues:
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: syntax -> approach?

File: proposals/0000-map-sorting.md [hunk 9]
```diff
@@ -34,17 +34,21 @@ people.sort { $0.age < $1.age }
...
+* The `$0.property < $1.property` syntax often leads to copy-and-paste bugs, since it requires duplicating the property accessor before and after the comparison operator.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: I think it's already clear enough why these expressions can cause copy-and-paste bugs.

File: proposals/0000-map-sorting.md [hunk 10]
```diff
@@ -34,17 +34,21 @@ people.sort { $0.age < $1.age }
...
+* For long property names, this syntax can be especially verbose since it requires repeating the property name twice.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: This is more of a severe case of the previous point. It was originally meant to highlight *non-trivial cases*, when computing values can take up a lot of lines.

File: proposals/0000-map-sorting.md [hunk 11]
```diff
@@ -8,13 +8,13 @@
...
 
-This proposal presents an addition to the Standard Library in an effort to make map sorting and, eventually, closureless key-path sorting a functional prerequisite for Swift.
+The authors propose adding an additional `sort` and `sorted` method on `MutableCollection` and `Sequence`, respectively, to improve the ergonomics for common use cases of the existing `sort` and `sorted` APIs.
 
 Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/t/map-sorting/21421)
 
 ## Motivation
 
-The straightforward way to sort a collection over a `Comparable` property of its `Element` type is to use a regular predicate, welding together the accessing and comparison of values.
+Today, the typical method for sorting arrays is to use an `areInIncreasingOrder` predicate. This closure takes the form `(Element, Element) -> Bool`, so it welds together both the accessing and comparison of values.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: I intentionally started here with the exact problem we are trying to solve - sorting over element metrics. The objective is to show why the regular predicate is unsuitable for *map sorting*.  The examples and disadvantages are written in the context of that problem and make less sense with a more general formulation. What is the motivation for being tight-lipped before the «Proposed Solution» section?

**Review by AnthonyLatsis — COMMENTED** [review]
Thanks Cal! Here are some comments.

File: proposals/0000-map-sorting.md (hunk 1 above)
Comment (calda) [reply proposals/0000-map-sorting.md:129]: I mean "default" in that you could still provide an alternative `areInIncreasingOrder` predicate. The implementation may be split across two functions in different `extension` contexts, but from a usage standpoint I think it's better to think of it as the default value.

File: proposals/0000-map-sorting.md (hunk 2 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Good catch!

File: proposals/0000-map-sorting.md (hunk 3 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Good point!

File: proposals/0000-map-sorting.md (hunk 5 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Ah, good point.

File: proposals/0000-map-sorting.md (hunk 7 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: I cut it because I find the phrase "a functional prerequisite for Swift" confusing -- what did you mean by that? *"prerequisite for Swift"* doesn't seem quite right, since this is a piece *of* Swift.

File: proposals/0000-map-sorting.md (hunk 10 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: What makes `verbose` different from `non-trivial`? `verbose` seems more clear, if we're talking about the number of lines or number of characters. 

I felt like `non-trivial` could have implied "computationally expensive".

File: proposals/0000-map-sorting.md (hunk 11 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Ah I see, I misinterpreted your original phrasing. I've updated this section:

> To sort a `Collection`, you're currently required to specify an `areInIncreasingOrder` predicate of the form `(Element, Element) -> Bool`. If you're sorting on the `Element` itself, you can use simple comparators like `<` and `>`.
> 
> Sorting a `Collection` on some *property* or *value* derived from the `Element` is more complex, and currently requires specifying a closure that welds together both the accessing and comparison of values.
>
> ...

**Comment by calda** [comment]
Thanks for the good feedback! For each of your points, I've gone back and either updated the proposal or provided additional rationale.

File: proposals/0000-map-sorting.md (hunk 6 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: I think it's worth mentioning that the `areInIncreasingOrder` predicate takes the form `(Value, Value) -> Bool`, since this is notably different from (although similar to) the existing `(Element, Element) -> Bool` form.

File: proposals/0000-map-sorting.md (hunk 10 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: verbose is fine, I was pointing at «For long property names».

File: proposals/0000-map-sorting.md (hunk 7 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: For Swift as a language. The standard library is written for the language, although it can also be called part of the tooling that makes up the language. 

I'm not against rephrasing that last part. The main point was to be a bit less opaque about what this proposal aims for.

File: proposals/0000-map-sorting.md (hunk 10 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Ah, understood.

File: proposals/0000-map-sorting.md (hunk 7 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Understood — thoughts on this?::

> The authors propose adding additional `sorted` and `sort` methods on `Sequence` and `MutableCollection`, respectively, to allow the caller to transform the `Collection`'s `Element` to some other `Value` and then sort by some comparison of those mapped values. This will improve the ergonomics for common use cases of the existing `sort` and `sorted` APIs.

File: proposals/0000-map-sorting.md [hunk 12]
```diff
@@ -148,10 +148,20 @@ This follows the precedent set by the existing `sort(by:)` and `sorted(by:)` met
...
 
 `people.sort(by: <) { $0.age }`
 
-During the pitch process, the authors concluded that `transform` should be positioned *before* the `areInIncreasingOrder` predicate. This better mirrors the flow of data, where the `transform` closure is always called before the `areInIncreasingOrder` predicate  `(Element) -> (Value), (Value, Value) -> (Bool)`.
+The authors concluded that `transform` should be positioned *before* the `areInIncreasingOrder` predicate. This better mirrors the flow of data, where the `transform` closure is always called before the `areInIncreasingOrder` predicate  `(Element) -> (Value), (Value, Value) -> (Bool)`.
+
+### `isExpensiveTransform`
+
+The `isExpensiveTransform` is an optional flag included in the proposed method so that the caller my opt-in to using the [Schwartzian Transform](https://en.wikipedia.org/wiki/Schwartzian_transform). The authors think that this optimization is useful enough to be worth including as a part of the proposal. Since it is an optional parameter (defaulting to `false`), it could alternatively be excluded. Callers seeking this sort of optimization would otherwise have to use a more complex pattern:
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: * the caller *may* opt-in
* to use a more complex *and generally less efficient* pattern (I think it is important to show that `isExpensiveTransform` is not just a matter of sugar).

Comment (calda) [reply proposals/0000-map-sorting.md]: Good point! Updated.

File: proposals/0000-map-sorting.md [hunk 13]
```diff
@@ -133,11 +133,11 @@ people.sort(on: { $0.age })
 people.sort(on: \.age)
 ```
 
-This follows the precedent set by the existing `sort(by:)` and `sorted(by:)` methods. The authors chose to exclude this additional overload from this initial proposal, as it is purely additive and could be included in either a later proposal or a final stage of this proposal (pending feedback during the review process).
+This follows the precedent set by the existing `sort(by:)` and `sorted(by:)` methods. The authors chose to exclude these additional overloads from this initial proposal, as they are purely additive and could be included in either a later proposal or a final stage of this proposal (pending feedback during the review process).
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: While we're here, let's refer to us as "We" for a change :)

File: proposals/0000-map-sorting.md [hunk 14]
```diff
@@ -160,7 +160,7 @@ The authors concluded that `transform` should be positioned *before* the `areInI
 
 ### `isExpensiveTransform`
 
-The `isExpensiveTransform` is an optional flag included in the proposed method so that the caller my opt-in to using the [Schwartzian Transform](https://en.wikipedia.org/wiki/Schwartzian_transform). The authors think that this optimization is useful enough to be worth including as a part of the proposal. Since it is an optional parameter (defaulting to `false`), it could alternatively be excluded. Callers seeking this sort of optimization would otherwise have to use a more complex pattern:
+The `isExpensiveTransform` is an optional flag included in the proposed method so that the caller may opt-in to using the [Schwartzian Transform](https://en.wikipedia.org/wiki/Schwartzian_transform). The authors think that this optimization is useful enough to be worth including as a part of the proposal. Since it is an optional parameter (defaulting to `false`), it could alternatively be excluded. Callers seeking this sort of optimization would otherwise have to use a more complex pattern that is both easy to get wrong and generally less efficient that the implementation provided in this proposal.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: generally less efficient *than* the implementation provided in this proposal?

Comment (calda) [reply proposals/0000-map-sorting.md]: 🤦🏻‍♂️ fixed

File: proposals/0000-map-sorting.md [hunk 15]
```diff
@@ -38,7 +38,7 @@ chats.sort { $0.lastMsg.date > $1.lastMsg.date }
 
 In many circumstances, this approach can cause issues:
 * The `$0.property < $1.property` syntax often leads to copy-and-paste bugs.
-* For long property names, this syntax can be especially verbose since it requires repeating the property name twice.
+* For long property names or complicated multi-line predicates, this syntax can be especially verbose since it requires repeating the value accessor twice.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: value accessor  -> calculations/computations?

Comment (calda) [reply proposals/0000-map-sorting.md]: Perhaps `since it requires repeating the accessor or computation twice`?

File: proposals/0000-map-sorting.md (hunk 11 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: But why so verbose? I would expect readers to be familiar with the Standard Library sorting APIs. Can we instead just say:

The straightforward/obvious way to sort a sequence over some property or value derived from its `Element` type is to use a regular predicate, welding together the accessing and comparison of values?

File: proposals/0000-map-sorting.md (hunk 15 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: I think "member" is more accurate in this case. We can also say 'retrieval' instead of separating chains and computations.

File: proposals/0000-map-sorting.md (hunk 11 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: It's better to be more verbose and more clear than less verbose and potentially less clear. I think that it's useful to explicitly lay out the status quo before explaining why the status quo has issues.

It wasn't immediately obvious to me upon first-read what a "regular predicate" was -- and we've been talking through this stuff for quite a while already. I think re-explaining the current state of the world is a good way to slowly ease the reader in to the terminology we use throughout this proposal (terminology which some readers may be unfamiliar with).

File: proposals/0000-map-sorting.md (hunk 15 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: `since it requires repeating the retrieval or computation twice`?

File: proposals/0000-map-sorting.md (hunk 7 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: I feel like it has become too wordy. "adding additional" doesn't sound too good either. We are about to explain everything in the motivation section; can we be more terse at least here?:

This proposal presents an addition to the Standard Library in an effort to make sorting over mapped values and, in particular, closureless key-path sorting an out-of-the-box convenience in Swift.

File: proposals/0000-map-sorting.md (hunk 15 above)
Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: "Since it requires repeating the retrieval twice". Or "since it requires duplicating the logic for retrieving the values". At your discretion.

File: proposals/0000-map-sorting.md (hunk 7 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: Perhaps:

> This proposal presents an addition to the Standard Library that makes it easy to sort a collection on some set of mapped values, provided via a `transform` closure or KeyPath, in a way that is ergonomic, idiomatic, and performant.

(I think we should stray away from the term "convenience", since we don't want this to get misconstrued as just syntactic sugar.

File: proposals/0000-map-sorting.md (hunk 15 above)
Comment (calda) [reply proposals/0000-map-sorting.md]: I like "since it requires duplicating the logic for retrieving the values"!

File: proposals/0000-map-sorting.md [hunk 16]
```diff
@@ -8,7 +8,7 @@
 
 ## Introduction
 
-The authors propose adding additional `sorted` and `sort` methods on `Sequence` and `MutableCollection`, respectively, to allow the caller to transform the `Collection`'s `Element` to some other `Value` and then sort by some comparison of those mapped values. This will improve the ergonomics for common use cases of the existing `sort` and `sorted` APIs.
+This proposal presents an addition to the Standard Library that makes it easy to sort a collection on some set of mapped values, provided via a transform closure or KeyPath, in a way that is ergonomic, idiomatic, and performant.
```
Comment (AnthonyLatsis) [inline proposals/0000-map-sorting.md]: One last nit: Please tick KeyPath if you were referring to the type.

Is "sort on a set" semantically correct? I am not quite a native speaker, but was under the impression "over" sounds better.

Comment (calda) [reply proposals/0000-map-sorting.md]: I used `sort on` since that mirrors the `sort(on:)` spelling that we're using in the implementation.

Comment (AnthonyLatsis) [reply proposals/0000-map-sorting.md]: It is merely an argument label. Besides, the target object has a different nature here. My gut tells me "Sort on a metric, but sort over a set".

Comment (calda) [reply proposals/0000-map-sorting.md]: Sure, sounds reasonable to me.

**Comment by AnthonyLatsis** [comment]
Thanks for being patient! Do you mind if I squash?

**Comment by calda** [comment]
sgtm, thanks for hashing everything out. It's a pretty good-looking proposal now 😄

**Comment by calda** [comment]
@AnthonyLatsis can you go back and update the [pitch thread](https://forums.swift.org/t/map-sorting/21421/)? It may also be nice to drop a note in [apple/swift-evolution#999](https://github.com/apple/swift-evolution/pull/999)
